tb_pulumi/ec2.py

```diff
@@ -0,0 +1,356 @@
+            disable_api_termination=False,  # be the source of data; they don't need protection.
+            instance_type='t3.micro',
+            key_name=self.resources['keypair']['keypair'].key_name,
+            root_block_device={'encrypted': True, 'kms_key_id': kms_key_id, 'volume_size': 10, 'volume_type': 'gp3'},
+            subnet_id=subnet_id,
+            user_data=user_data,
+            volume_tags=self.tags,
+            vpc_security_group_ids=sg_ids,
+            tags=instance_tags,
+            opts=pulumi.ResourceOptions(parent=self),
+        )
+
+        self.finish()
+
+
+class SshKeyPair(tb_pulumi.ThunderbirdComponentResource):
+    """Builds an SSH keypair and stores its values in Secrets Manager.
+
+    NOTE: This should typically be used by specifying the public_key. If you do not, Pulumi will
+    generate a new key for you. However, at the moment, it appears there's no way to have Pulumi
+    generate a private key ONE TIME and ONLY ONE TIME. Each `pulumi up/preview` command generates a
+    new keypair, which generates new secret versions (and if this is attached to an instance
+    downstream, it triggers the recreation of that instance). This is otherwise good code that will
+    correctly build these resources.
+    """
+
+    def __init__(
+        self,
+        name: str,
+        project: tb_pulumi.ThunderbirdPulumiProject,
+        key_size: int = 4096,
+        public_key: str = None,
+        secret_name: str = 'keypair',
+        opts: pulumi.ResourceOptions = None,
+        **kwargs,
+    ):
+        """Construct an SshKeyPair.
+
+        Positional arguments:
+            - name: A string identifying this set of resources.
+            - project: The ThunderbirdPulumiProject to add these resources to.
+
+        Keyword arguments:
+            - key_size: Byte length of the private key to generate. Only used if public_key is not
+                supplied.
+            - public_key: RSA public key to stash in the KeyPair. It is highly recommended that you
+                always provide this. That is, you should usually generate a keypair on your local
+                machine (ssh-keygen -t rsa -b 4096) and provide that public key to this resource.
+            - secret_name: A slash ("/") delimited name to give the Secrets Manager secret. If not
+                supplied, one will be generated based on `name`. Only used if public_key is not
+                provided.
+            - opts: Additional pulumi.ResourceOptions to apply to these resources.
+            - kwargs: Any other keyword arguments which will be passed as inputs to the
+                ThunderbirdComponentResource superconstructor.
+        """
+
+        super().__init__('tb:ec2:SshKeyPair', name, project, opts=opts, **kwargs)
+
+        if not public_key:
+            self.resources['private_key'], self.resources['public_key'] = generate_ssh_keypair(key_size=key_size)
```

Comment:
Should we log about the fact we're making a new keypair?

Reply:
It becomes very obvious in the `pulumi preview` output that you're building a new keypair and replacing other things that depend on it.